### DISCUSSION TITLE
Recruiting script authors, and public beta?

### DISCUSSION THREAD

**Discussion by artillect** [aclist/kbin-megamod#30] [Archived]
I think we're getting pretty close to a public release! It just seems to me that we need to finalize the config API (by sending the object with the options packed into it as discussed in #8), get some scripts ported/written, and sort out some of the issues, and we're ready to go. I'm not sure if we want to recruit more script authors before the config API is finalized, but we could make a post on /m/kbinstyles asking people to port their scripts and letting them know about /m/enhancement. (It could be nice to get more contributors to the core project as well to help out with the various option types.)

**Comment by aclist** [comment]
I'd rather not jump the gun until the API is rock solid, as it makes or breaks adoption. It should be as easy as possible to just integrate a script by adding a few lines (best case). If we have to make posts about this repeatedly in a short span of time, it's liable to be annoying. 

I think users kind of are in a "public beta" already when it comes to using self-installed scripts, so better to iron out kinks here so that it's quite functional end to end. In particular I'm thinking of the need to handle the case when the remote data is unretrievable or the resource URL is down, and filling the modal with some boilerplate warning rather than just erroring out and coming up blank. Or maybe you are thinking about first mover advantage here?

I am dragging my heels with the API because I was initially going to send both the toggle and the options as an object and have the recipient script process this as an argument upfront. But while looking at some other scripts, I noticed that some use a similar config storage logic to megamod and namespace their config settings with a prefix (e.g. "myscriptsettings.setting1"). So this seems like a pretty good standard to use, where if the recipient script has persistent options, it can look to this named data. In other words, I don't think the script entrypoint needs to read arguments other than toggle on/off. After that, it can check its settings file in whatever way it wants to.

- Bind an event listener to the helpbox and use event delegation. Click events are not mapped literally to each field but generically to the helpbox, so we only have one listener. If a click event target matches any of checkbox, selector, textarea, or colorpicker, store its value in a separate, namespaced localstorage object
- Store the toggle state in megamod's own settings object
- When setting up input fields a priori in the manifest.json, set the namespace and keys that the recipient script expects to read its configs from
- Helpbox events update into these keys
- When options are changed, call the script again, unless the state is toggled off already

So, wrapping that all up:

*MYSCRIPT.js*:
- Expects localstorage data prefixed with the name 'myscript'
- Expects the keys `color` and `text` and exposes a colorpicker and textarea

*manifest.json*:

```
  {
    "name": "MYSCRIPT",
    "author": "foo",
    "version": "0.1.0",
    "label": "Action",
    "desc": "Details",
    "link": "",
    "linkLabel": "",
    "class": "init",
    "entrypoint": "init",
    "settings": "myscript" <--- name of the localstorage 
    "fields": [                           <--- array of input fields
                  {
                 "type": "colorpicker",
                  "initial": "#000000",
                   "key": "color",  <---  recipient script expects 'myscript.color' to contain the the hex value
                  }
                  {
                 "type": "textarea",
                  "initial": "enter text here"
                   "key": "text",   <---- recipient script expects 'myscript.text' to contain the raw text
                  }
       ]
    "page": "general"
  },
```

Functional benefit of this: recipient scripts already look for some settings parameters with specific names. Rather than forcing a new convention on them and arbitrarily deciding on our own config scheme here and stuffing the entry point function with that, use the literal keys set up by the script and write into those. Then the recipient script just gets toggled on/off and proceeds to read localstorage as it normally would within its own logic.

It makes the manifest more verbose upfront, but I think this is good and everything is clearly delineated. You can see what the fields are and how they function at the manifest level. It should make the porting logic considerably simpler, since everything can be identical to the recipient script, and we just need to watch out for collisions and possibly encourage adoption of proper namespacing (`SCRIPTNAME.setting` in the localstorage).

**Reply by artillect** [reply]
> I'd rather not jump the gun until the API is rock solid, as it makes or breaks adoption. It should be as easy as possible to just integrate a script by adding a few lines (best case). If we have to make posts about this repeatedly in a short span of time, it's liable to be annoying.

That's a good point, it'd annoy everyone to port their scripts over and then have to update them all again. Once the API is solidified I think it'd be good to get a post out. Or maybe, just get everything but the extra config options finalized and then add them on top of that (if that's possible). That way, we could start having people port their scripts, and then add config options later without requiring other changes.

> I think users kind of are in a "public beta" already when it comes to using self-installed scripts, so better to iron out kinks here so that it's quite functional end to end. In particular I'm thinking of the need to handle the case when the remote data is unretrievable or the resource URL is down, and filling the modal with some boilerplate warning rather than just erroring out and coming up blank.

I think that's all a good plan, I just think that, given that using userscripts is basically already a public beta, people aren't really going to mind some of the rough edges as much, and I feel like it's easier to make adjustments early. It'd be good to get feedback from more people so we don't start painting ourselves into a corner.

> Or maybe you are thinking about first mover advantage here?

Exactly, I wanna try to give script authors a single banner to rally under a single banner so there isn't as much fragmentation, and so authors of different scripts don't have to reinvent the wheel for their script collection with a config menu.

I think your final plan for the API sounds good, but I think I'm a little confused on how the config options get handled. Let me know if I'm understanding this correctly:

- Script author adds fields to manifest
- Config modal populates helpbox with additional options and stores values in `[mod].[field]`s
- Mod gets loaded as normal via the entrypoint with only the toggle being passed
- Mod loads values of config from `[mod].[field]`s (? This is the step I'm confused about. Is the loading from local storage more involved than that?)

**Reply by aclist** [reply]
Alright, I'll make a to-do list of what I consider minimum tasks before releasing 1.0.0. Outside of API, mostly administrative:

- [x] Config API
- [x] Fix the bug mentioned in #26 and appearance of the button (assignee: artillect) 
  - While you are it, seems at some point we lost the hover color when rolling over links in the list. Can you restore it?
    - Preferably rolling over the ACTIVE page or mod will not change its color
- [x] Create an issue template for the bug tracker (assignee: aclist)
- [ ] Merge and integration test minimum collection of scripts (to show basic functionality)
  - We should have enough by now. Just need to audit each one and update to the new API logic (assignee: all)
  - Most of my mods are broken and not loading after the DOM
- [ ] Finalize working name (assignee: all) 
  - Cf. https://github.com/aclist/kbin-megamod/discussions/15
- [ ] Update documentation describing new manifest scheme (assignee: aclist)
- [ ] Revamp #12 (assignee: aclist; may want help)

> I think your final plan for the API sounds good, but I think I'm a little confused on how the config options get handled. Let me know if I'm understanding this correctly:

You understood it all correctly. If a script has persistent settings, presumably they have to be stored somewhere, and this should keep megamod as generic as possible in that it merely writes changes to wherever the recipient script expects, rather than trying to load the script and saying "we want to make this change, and this change, and this change," and the script having to be refactored to parse all of that upfront, instead of just starting up and doing whatever it would normally do. 

I think it makes sense to enforce a reasonable standard here. They can do whatever they want with the options, but in the interest of a script (could be any script other than megamod) accessing the options of another one, they should use a standardized format so that they are namespaced and accessible from localstorage. I think this is the only part of the logic it is worth being opinionated about. We can strongly suggest adhering to this format, and I think most scripts already do so. Worst case scenario, we can port their script and ensure it follows this rubric. I think this change is important not just for the sake of megamod, but for script "IPC" more generally and represents a reasonable way forward.

By way of explanation, I will include a sample script when updating the API logic. I am open to changing this idea, but I believe it should keep everything sane(r).

**Reply by aclist** [reply]
As for adding multi-options first, I consider this to be a basic feature and expected by any user or author alike. We have the manifest schema and other areas built around this, so it makes sense to add it, because it's rather confusing and frustrating to expect to be able to add form fields and see that that part is missing. I'll be working on it later on, don't think it's going to eat up too much time.

**Reply by artillect** [reply]
Good point, it's a bit out of my wheelhouse so I wasn't sure how long it'd take

**Reply by aclist** [reply]
Like most things, it's 95% conceptualization and 5% implementation.

**Reply by aclist** [reply]
The new API logic is up and running. In principle, it accepts any of the input types defined [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input). Still have to add support for selectors (dropdowns), since they are a special case and not considered an input. This is completely generic and supports any input type without additional handling.

It should be possible to arbitrarily add fields in the manifest now and have them update in the helpbox. You must also give them a label (currently no error handling scriptside, but I plan to add an acceptance test on the JSON itself to validate fields. Regardless, we have to approve manifest changes anyway, so I can't see how malformed manifests would leak into the codebase.)

I added a mod called `yellow.js` to the mods directory which shows a proof of concept of reading from settings. It is set up in the manifest as well. It starts with a colorpicker set to yellow hex color and can be updated to any other. You should see changes propagate immediately, and they are persistent to that mod.

I also set up basic fields for the existing mods; feel free to correct them after testing.

**Reply by artillect** [reply]
Looks good to me, but I feel like `readSettings()` should get moved out of `yellow.js` and into `megamods.user.js` so it's part of the API. I think it might also cause collisions if each script author defines their own readSettings function. Maybe we need to shove each script into a class? Although I'm not sure how that'd affect the manifest/`funcObj`

**Reply by aclist** [reply]
Hmm, good point. This function is identical to one already used internally in megamod. So the recipient script can just pass its desired namespace to that function and parse the results as necessary. Easy fix and I think we can handle collisions in function names etc. through acceptance testing. The test didn't find any duplicates yet, but did reveal some redundant lines of code.

**Reply by aclist** [reply]
Radio and select type fields should be working now, along with some other general bugfixes to the API logic. We can start integrating mods and checking whether they get the desired input field values back. I've added logging to the console when an element is updated or toggled to print the event target and the current content of localStorage, which should make it clear what values are currently being retained.
